floss/identify.py

```diff
@@ -0,0 +1,130 @@
+
+
+def is_thunk_function(vw, function_address):
+    return vw.getFunctionMetaDict(function_address).get("Thunk", False)
+
+
+def get_function_score_weighted(features):
+    return sum(feature.weighted_score() for feature in features) / sum(feature.weight for feature in features)
+
+
+def find_decoding_functions(vw, functions, disable_progress=None):
+    decoding_candidate_functions = collections.defaultdict(float)
+
+    pbar = tqdm.tqdm
+    if disable_progress:
+        # do not use tqdm to avoid unnecessary side effects when caller intends
+        # to disable progress completely
+        pbar = lambda s, *args, **kwargs: s
+
+    # TODO see FLIRT below
+    # n_funcs = len(functions)
+
+    functions = sorted(functions)
+
+    pb = pbar(functions, desc="finding decoding functions", unit=" functions", postfix="skipped 0 library functions")
+    for f in pb:
+        function_address = int(f)
+
+        if is_thunk_function(vw, function_address):
+            continue
+
+        # TODO ignore library functions via FLIRT
+        # if viv_utils.flirt.is_library_function(vw, function_address):
+        #     function_name = viv_utils.get_function_name(vw, function_address)
+        #     logger.debug("skipping library function 0x%x (%s)", function_address, function_name)
+        #     meta["library_functions"][function_address] = function_name
+        #     n_libs = len(meta["library_functions"])
+        #     percentage = 100 * (n_libs / n_funcs)
+        #     if isinstance(pb, tqdm.tqdm):
+        #         pb.set_postfix_str("skipped %d library functions (%d%%)" % (n_libs, percentage))
+        #     continue
+
+        f = viv_utils.Function(vw, function_address)
+
+        function_data = {"meta": get_function_meta(f), "features": list()}
+
+        # meta data features
+        function_data["features"].append(BlockCount(function_data["meta"].get("block_count")))
+        function_data["features"].append(InstructionCount(function_data["meta"].get("instruction_count")))
+        function_data["features"].append(Arguments(function_data["meta"].get("api", []).get("arguments")))
+
+        for feature in extract_function_features(f):
+            function_data["features"].append(feature)
+
+        for bb in f.basic_blocks:
+            for feature in extract_basic_block_features(f, bb):
+                function_data["features"].append(feature)
+
+            for insn in bb.instructions:
+                for feature in extract_insn_features(f, bb, insn):
+                    function_data["features"].append(feature)
+
+        for feature in abstract_features(function_data["features"]):
+            function_data["features"].append(feature)
+
+        function_data["score"] = get_function_score_weighted(function_data["features"])
+
+        logger.debug("analyzed function 0x%x - total score: %f", function_address, function_data["score"])
+        for feat in function_data["features"]:
+            logger.debug("  %s", feat)
```

Comment:
example output

```
$ floss tests/data/src/decode-to-stack/bin/test-decode-to-stack.exe --no-static-strings --no-stack-strings -d
[...]
INFO: floss: identifying decoding functions...
DEBUG: floss.identify: analyzed function 0x401000 - total score: 0.100000
DEBUG: floss.identify:   BlockCount = 1 (score: 0.40, weighted: 0.10)
DEBUG: floss.identify:   InstructionCount = 1 (score: 0.10, weighted: 0.03)
DEBUG: floss.identify:   Arguments = 0 (score: 0.00, weighted: 0.00)
DEBUG: floss.identify:   CallsTo = 0 (score: 0.00, weighted: 0.00)
DEBUG: floss.identify: analyzed function 0x4014e0 - total score: 0.784211
DEBUG: floss.identify:   BlockCount = 42 (score: 0.10, weighted: 0.03)
DEBUG: floss.identify:   InstructionCount = 211 (score: 0.80, weighted: 0.20)
DEBUG: floss.identify:   Arguments = 0 (score: 0.00, weighted: 0.00)
DEBUG: floss.identify:   CallsTo = 0 (score: 0.00, weighted: 0.00)
DEBUG: floss.identify:   Loop = 38 (score: 1.00, weighted: 0.50)
DEBUG: floss.identify:   Nzxor = 0x4012ea  xor ebx,1 (score: 1.00, weighted: 0.75)
DEBUG: floss.identify:   TightLoop = (4199264, 4199319) (score: 1.00, weighted: 0.75)
DEBUG: floss.identify:   Mov = 0x4013bd  mov dword [eax],edx (score: 1.00, weighted: 0.50)
DEBUG: floss.identify:   NzxorLoop = True (score: 1.00, weighted: 1.00)
DEBUG: floss.identify: analyzed function 0x401500 - total score: 0.417143
DEBUG: floss.identify:   BlockCount = 7 (score: 1.00, weighted: 0.25)
DEBUG: floss.identify:   InstructionCount = 23 (score: 0.80, weighted: 0.20)
DEBUG: floss.identify:   Arguments = 0 (score: 0.00, weighted: 0.00)
DEBUG: floss.identify:   CallsTo = 2 (score: 0.14, weighted: 0.07)
DEBUG: floss.identify: analyzed function 0x401509 - total score: 0.360000
DEBUG: floss.identify:   BlockCount = 5 (score: 1.00, weighted: 0.25)
DEBUG: floss.identify:   InstructionCount = 20 (score: 0.80, weighted: 0.20)
DEBUG: floss.identify:   Arguments = 0 (score: 0.00, weighted: 0.00)
DEBUG: floss.identify:   CallsTo = 0 (score: 0.00, weighted: 0.00)
DEBUG: floss.identify: analyzed function 0x401550 - total score: 0.100000
DEBUG: floss.identify:   BlockCount = 1 (score: 0.40, weighted: 0.10)
DEBUG: floss.identify:   InstructionCount = 4 (score: 0.10, weighted: 0.03)
DEBUG: floss.identify:   Arguments = 0 (score: 0.00, weighted: 0.00)
DEBUG: floss.identify:   CallsTo = 0 (score: 0.00, weighted: 0.00)
DEBUG: floss.identify: analyzed function 0x401560 - total score: 0.827551
DEBUG: floss.identify:   BlockCount = 9 (score: 1.00, weighted: 0.25)
DEBUG: floss.identify:   InstructionCount = 50 (score: 0.80, weighted: 0.20)
DEBUG: floss.identify:   Arguments = 6 (score: 0.50, weighted: 0.12)
DEBUG: floss.identify:   CallsTo = 2 (score: 0.14, weighted: 0.07)
DEBUG: floss.identify:   Loop = 2 (score: 1.00, weighted: 0.50)
DEBUG: floss.identify:   Nzxor = 0x4015b4  xor eax,ecx (score: 1.00, weighted: 0.75)
DEBUG: floss.identify:   NzxorLoop = True (score: 1.00, weighted: 1.00)
DEBUG: floss.identify: analyzed function 0x4015f0 - total score: 0.617143
DEBUG: floss.identify:   BlockCount = 4 (score: 1.00, weighted: 0.25)
DEBUG: floss.identify:   InstructionCount = 65 (score: 0.80, weighted: 0.20)
DEBUG: floss.identify:   Arguments = 2 (score: 1.00, weighted: 0.25)
DEBUG: floss.identify:   CallsTo = 2 (score: 0.14, weighted: 0.07)
DEBUG: floss.identify: analyzed function 0x401740 - total score: 0.620408
DEBUG: floss.identify:   BlockCount = 12 (score: 0.40, weighted: 0.10)
DEBUG: floss.identify:   InstructionCount = 36 (score: 0.80, weighted: 0.20)
DEBUG: floss.identify:   Arguments = 3 (score: 1.00, weighted: 0.25)
DEBUG: floss.identify:   CallsTo = 1 (score: 0.07, weighted: 0.04)
DEBUG: floss.identify:   Loop = 3 (score: 1.00, weighted: 0.50)
DEBUG: floss.identify: analyzed function 0x4017e0 - total score: 0.471429
DEBUG: floss.identify:   BlockCount = 1 (score: 0.40, weighted: 0.10)
DEBUG: floss.identify:   InstructionCount = 2 (score: 0.10, weighted: 0.03)
DEBUG: floss.identify:   Arguments = 1 (score: 1.00, weighted: 0.25)
DEBUG: floss.identify:   CallsTo = 6 (score: 0.43, weighted: 0.21)
DEBUG: floss.identify: analyzed function 0x4017f0 - total score: 0.414286
DEBUG: floss.identify:   BlockCount = 1 (score: 0.40, weighted: 0.10)
DEBUG: floss.identify:   InstructionCount = 2 (score: 0.10, weighted: 0.03)
DEBUG: floss.identify:   Arguments = 1 (score: 1.00, weighted: 0.25)
DEBUG: floss.identify:   CallsTo = 4 (score: 0.29, weighted: 0.14)
DEBUG: floss.identify: analyzed function 0x401800 - total score: 0.617143
DEBUG: floss.identify:   BlockCount = 3 (score: 1.00, weighted: 0.25)
DEBUG: floss.identify:   InstructionCount = 46 (score: 0.80, weighted: 0.20)
DEBUG: floss.identify:   Arguments = 1 (score: 1.00, weighted: 0.25)
DEBUG: floss.identify:   CallsTo = 2 (score: 0.14, weighted: 0.07)
DEBUG: floss.identify: analyzed function 0x4018c0 - total score: 0.414286
DEBUG: floss.identify:   BlockCount = 1 (score: 0.40, weighted: 0.10)
DEBUG: floss.identify:   InstructionCount = 10 (score: 0.10, weighted: 0.03)
DEBUG: floss.identify:   Arguments = 1 (score: 1.00, weighted: 0.25)
DEBUG: floss.identify:   CallsTo = 4 (score: 0.29, weighted: 0.14)
DEBUG: floss.identify: analyzed function 0x4018e0 - total score: 0.440000
DEBUG: floss.identify:   BlockCount = 26 (score: 0.40, weighted: 0.10)
DEBUG: floss.identify:   InstructionCount = 84 (score: 0.80, weighted: 0.20)
DEBUG: floss.identify:   Arguments = 1 (score: 1.00, weighted: 0.25)
DEBUG: floss.identify:   CallsTo = 0 (score: 0.00, weighted: 0.00)
DEBUG: floss.identify: analyzed function 0x401b00 - total score: 0.942105
DEBUG: floss.identify:   BlockCount = 63 (score: 0.10, weighted: 0.03)
DEBUG: floss.identify:   InstructionCount = 367 (score: 0.80, weighted: 0.20)
DEBUG: floss.identify:   Arguments = 3 (score: 1.00, weighted: 0.25)
DEBUG: floss.identify:   CallsTo = 14 (score: 1.00, weighted: 0.50)
DEBUG: floss.identify:   Loop = 4 (score: 1.00, weighted: 0.50)
DEBUG: floss.identify:   Loop = 7 (score: 1.00, weighted: 0.50)
DEBUG: floss.identify:   Loop = 7 (score: 1.00, weighted: 0.50)
DEBUG: floss.identify:   Loop = 3 (score: 1.00, weighted: 0.50)
DEBUG: floss.identify:   Shift = 0x401bbb  shl ecx,2 (score: 1.00, weighted: 0.75)
DEBUG: floss.identify:   TightLoop = (4202176, 4202221) (score: 1.00, weighted: 0.75)
DEBUG: floss.identify: analyzed function 0x401b60 - total score: 0.904511
DEBUG: floss.identify:   BlockCount = 62 (score: 0.10, weighted: 0.03)
DEBUG: floss.identify:   InstructionCount = 348 (score: 0.80, weighted: 0.20)
DEBUG: floss.identify:   Arguments = 3 (score: 1.00, weighted: 0.25)
DEBUG: floss.identify:   CallsTo = 9 (score: 0.64, weighted: 0.32)
DEBUG: floss.identify:   Loop = 4 (score: 1.00, weighted: 0.50)
DEBUG: floss.identify:   Loop = 7 (score: 1.00, weighted: 0.50)
DEBUG: floss.identify:   Loop = 7 (score: 1.00, weighted: 0.50)
DEBUG: floss.identify:   Loop = 3 (score: 1.00, weighted: 0.50)
DEBUG: floss.identify:   Shift = 0x401bbb  shl ecx,2 (score: 1.00, weighted: 0.75)
DEBUG: floss.identify:   TightLoop = (4202176, 4202221) (score: 1.00, weighted: 0.75)
DEBUG: floss.identify: analyzed function 0x401da0 - total score: 0.682143
DEBUG: floss.identify:   BlockCount = 37 (score: 0.10, weighted: 0.03)
DEBUG: floss.identify:   InstructionCount = 188 (score: 0.80, weighted: 0.20)
DEBUG: floss.identify:   Arguments = 0 (score: 0.00, weighted: 0.00)
DEBUG: floss.identify:   CallsTo = 2 (score: 0.14, weighted: 0.07)
DEBUG: floss.identify:   Loop = 4 (score: 1.00, weighted: 0.50)
DEBUG: floss.identify:   Loop = 7 (score: 1.00, weighted: 0.50)
DEBUG: floss.identify:   TightLoop = (4202176, 4202221) (score: 1.00, weighted: 0.75)
[...]
```

Reply:
beautiful! and easy as a human to interpret.

only recommendation might be to justify the feature names (`name.ljust(32)`) so that the numbers line up. low priority.